src/main/scala/chisel3/util/Decoupled.scala

```diff
@@ -227,20 +232,34 @@ class Queue[T <: Data](val gen: T,
   val do_deq = WireDefault(io.deq.fire())
 
   when (do_enq) {
-    ram(enq_ptr.value) := io.enq.bits
+    if (useSyncReadMem) {
+      ram.write(enq_ptr.value, io.enq.bits)
+    }
+    else {
+      ram(enq_ptr.value) := io.enq.bits
+    }
     enq_ptr.inc()
   }
   when (do_deq) {
     deq_ptr.inc()
+    if (useSyncReadMem)
+      deq_ptr_value.get := Mux(deq_ptr.value === (entries.U - 1.U), 0.U, deq_ptr.value + 1.U)
```

Comment:
As an unfortunate consequence of how the behavioral `read` method works (the read is enabled when the address is updated), this actually uses `do_deq` as the read enable.

Reply:
I guess you wanted to say that in this case the `read` method with enable signal can be potentially used and so the code could be organized differently. However, it seems that the `val ram = if ...` assignment actually causes that specific `read` method with enable signal is not visible and hence can not be used due compile-time error.